lib/kafkat/command/set-replication-factor.rb

```diff
@@ -0,0 +1,173 @@
+module Kafkat
+  module Command
+
+    #
+    # Command to set the replication factor (RF) of a topic.
+    # The command accepts the topic name, the new desired replication factor,
+    # and, in case of an increase of the replication factor, a list of broker ids.
+    #
+    # When reducing the RF, a new partition assigment will be generated by
+    # removing the last partition replica of every replica set. The leader partition
+    # will not be removed from the replica set to prevent a leader election.
+    #
+    # When increasing the RF, a new partition assignment will be generated by allocating
+    # a new replica to every replica set. The new replica will be assigned to the provided
+    # broker list in a round robin fashion.
+    # If no broker id is specified on the command line, all brokers will be used.
+    #
+    #
+    class SetReplicationFactor < Base
+      register_as 'set-replication-factor'
+
+      usage 'set-replication-factor [topic] [--replicas <n>]',
+            'Set the replication factor of'
+
+      def run
+        topic_name = ARGV.shift unless ARGV[0] && ARGV[0].start_with?('--')
+
+        all_brokers = zookeeper.get_brokers
+        topics = topic_name && zookeeper.get_topics([topic_name])
+        topics ||= zookeeper.get_topics
+
+        opts = Trollop.options do
+          opt :brokers, "the comma-separated list of broker the new partitions must be assigned to", type: :string
+          opt :newrf, "the new replication factor", type: :integer, required: true
+        end
+
+        broker_ids = opts[:brokers] && opts[:brokers].split(',').map(&:to_i)
+        new_rf = opts[:newrf]
+
+        if new_rf.nil? || new_rf < 1
+          puts "ERROR: replication factor is smaller than 1"
+          exit 1
+        end
+
+        broker_ids ||= zookeeper.get_brokers.values.map(&:id)
+
+        all_brokers_id = all_brokers.values.map(&:id)
+        broker_ids.each do |id|
+          if !all_brokers_id.include?(id)
+            puts "ERROR: Broker #{id} is not currently active.\n"
+            exit 1
+          end
+        end
+
+        broker_count = broker_ids.size
+        if new_rf > broker_count
+          puts "ERROR: Replication factor is larger than number of brokers.\n"
+          exit 1
+        end
+
+        assignments = []
+        topics.each do |_, t|
+          current_rf = t.partitions[0].replicas.size
+          topic_assignments = if new_rf < current_rf
```

Comment:
A minor detail, but I had trouble scanning this section. Ruby's "all statements are expressions" philosophy is (in general) a tough sell for me (likewise, with Scala). Since `reduce_rf` and `increase_rf` never return `nil`, it'd be much clearer if this transformation read like this —

```
def run

  ...

  assignments = []
  topics.each do |_, t|
    current_rf = t.partitions[0].replicas.size
    if new_rf < current_rf
      warn_reduce_brokers if opts[:brokers]
      assignments += reduce_rf(t, current_rf, new_rf)
    elsif new_rf > current_rf
      assignments += increase_rf(t, current_rf, new_rf, broker_ids)
    end
  end

  ...

end

def warn_reduce_brokers
  return if @did_warn_reduce_brokers
  puts "When reducing the replication factor the list of specified brokers is ignored."
  puts "Once the replication factor is set, you can use the reassign command."
  @did_warn_reduce_brokers = true
end
```

What do you think?